learn2track/models/gru_mixture.py

```diff
@@ -276,14 +280,65 @@ def _compute_losses(self, model_output):
         # means.shape : (batch_size, seq_len, n_gaussians, 3)
 
         # expected_value.shape : (batch_size, seq_len, 3)
-        expected_value = T.sum(mixture_weights[:, :, :, None] * means, axis=2)
+        self.samples = T.sum(mixture_weights[:, :, :, None] * means, axis=2)
+
+        # loss_per_time_step.shape = (batch_size, seq_len)
+        self.loss_per_time_step = T.sqrt(T.sum(((self.samples - self.dataset.symb_targets)**2), axis=2))
+        # loss_per_seq.shape = (batch_size,)
+        self.loss_per_seq = T.sum(self.loss_per_time_step*mask, axis=1) / T.sum(mask, axis=1)
+
+        return self.loss_per_seq
+
+    def _get_mixture_parameters(self, regression_output):
+        batch_size, seq_len = regression_output.shape[:2]
+
+        gaussian_3d_params_shape = (batch_size, seq_len, self.n, 3)
+
+        mixture_weights = softmax(regression_output[:, :, :self.n], axis=2)
+        means = T.reshape(regression_output[:, :, self.n:self.n * 4], gaussian_3d_params_shape)
+        stds = T.reshape(T.exp(regression_output[:, :, self.n * 4:self.n * 7]), gaussian_3d_params_shape)
+
+        return mixture_weights, means, stds
 
-        # L2_errors_per_time_step.shape = (batch_size, seq_len)
-        self.L2_errors_per_time_step = T.sqrt(T.sum(((expected_value - self.dataset.symb_targets)**2), axis=2))
-        # avg_L2_error_per_seq.shape = (batch_size,)
-        self.avg_L2_error_per_seq = T.sum(self.L2_errors_per_time_step*mask, axis=1) / T.sum(mask, axis=1)
 
-        return self.avg_L2_error_per_seq
+class MultivariateGaussianMixtureMaxComponentL2Distance(Loss):
+    """ Computes the L2 distance for the max component samples of a multivariate gaussian mixture
+    """
+
+    def __init__(self, model, dataset):
+        super().__init__(model, dataset)
+        self.n = model.n_gaussians
+        self.d = model.output_size
+
+    def _get_updates(self):
+        return {}  # There is no updates for L2Distance.
+
+    def _compute_losses(self, model_output):
+        mask = self.dataset.symb_mask
+
+        # regression_outputs.shape = (batch_size, seq_length, regression_layer_size)
+        regression_outputs = model_output
+
+        # mixture_weights.shape : (batch_size, seq_len, n_gaussians)
+        # means.shape : (batch_size, seq_len, n_gaussians, 3)
+        mixture_weights, means, stds = self._get_mixture_parameters(regression_outputs)
+
+        batch_size = mixture_weights.shape[0]
+        seq_len = mixture_weights.shape[1]
+
+        row_ids = T.tile(T.arange(0, batch_size), (seq_len,1)).T    # [[0,0,...,0],[1,1,...,1],...]
```

Comment:
Neat. But I **think** you don't need the `T.tile` and just rely on broadcasting. I might be wrong, though. Something, along the lines of
```python
    # Compute NLLs
    # predictions.shape: (batch_size, seq_length, out_size)
    neg_log_probs = -T.log(predictions)
    nlls = neg_log_probs[T.arange(class_ids.shape[0])[:, None],
                         T.arange(class_ids.shape[1])[None, :],
                         class_ids[:, :, 0]]
```